system-addon/lib/TopSitesFeed.jsm

```diff
@@ -154,39 +154,66 @@ this.TopSitesFeed = class TopSitesFeed {
       return pinnedLink;
     });
   }
+
+  /**
+   * Inform others that top sites data has been updated due to pinned changes.
+   */
   _broadcastPinnedSitesUpdated() {
-    this.store.dispatch(ac.BroadcastToContent({
-      type: at.PINNED_SITES_UPDATED,
-      data: this._getPinnedWithData()
-    }));
+    // Refresh to update pinned sites with screenshots, trigger deduping, etc.
+    this.refresh();
```

Comment:
The first pinning implementation called `refresh()` on every pinning change. Everything seems to come back around, heh. When we get drag and drop that might cause a lot of calls in a short period of time. But I'm assuming refresh will now used cached places db results?

Reply:
Hmm.. For drag-n-drop I would think the page is locally updating state during the drag until a drop to actually update the pinned states for things, and that should only trigger one `refresh`?